content/posts/csv-boilerplate-for-go.md

```diff
@@ -0,0 +1,307 @@
+
+To begin, I decided I wanted to use a real world CSV file as my example that has a bit of the typical uglyness one has to deal with. I quickly [found](https://perso.telecom-paristech.fr/eagan/class/igr204/datasets) a file called [cars.csv](https://github.com/felixge/dump/blob/master/csv-boilerplate/cars.csv) which looks like this:
+
+```csv
+Car;MPG;Cylinders;Displacement;Horsepower;Weight;Acceleration;Model;Origin
+STRING;DOUBLE;INT;DOUBLE;DOUBLE;DOUBLE;DOUBLE;INT;CAT
+Chevrolet Chevelle Malibu;18.0;8;307.0;130.0;3504.;12.0;70;US
+Buick Skylark 320;15.0;8;350.0;165.0;3693.;11.5;70;US
+Plymouth Satellite;18.0;8;318.0;150.0;3436.;11.0;70;US
+...
+```
+
+As you can see, it uses semicolons instead of commas as a separator, and it has a second header line that seems to contain type information for whatever reason. Not great, not terrible, but pretty typical for what's out there.
+
+The goals for my boilerplate is to read and write the file format in a way that solves the three problems outlined in the begining. It should do so without an excessive amount of abstraction, and certainly no reflection, yet as elegantly as possible. There should also be an intermediate Go struct to hold the data in memory as shown below.
+
+```go
+type Car struct {
+	Car          string
+	MPG          float64
+	Cylinders    int64
+	Displacement float64
+	Horsepower   float64
+	Weight       float64
+	Acceleration float64
+	Model        int64
+	Origin       string
+}
+```
+
+My first naive attempt is usually to write some code that defines the headers and struct mapping in a way that is separate from the CSV encoding/decoding itself. An example of that can be seen below:
+
+```go
+var carHeaders = []string{
+  "Car",
+  "MPG",
+  "Cylinders",
+  // remaining columns ...
+}
+
+func (c *Car) UnmarshalRecord(record []string) (err error) {
+	if got, want := len(record), len(carHeaders); got != want {
+		return fmt.Errorf("bad column number: got=%d want=%d", got, want)
+	}
+	c.Car = record[0]
+	if c.MPG, err = strconv.ParseFloat(record[1], 64); err != nil {
+		return fmt.Errorf("column=%q: %w", carHeaders[1], err)
+	} else if c.Cylinders, err = strconv.ParseInt(record[2], 10, 64); err != nil {
+		return fmt.Errorf("column=%q: %w", carHeaders[2], err)
+	}
+	// remaining columns ...
+	return nil
+}
+
+func (c *Car) MarshalRecord() ([]string, error) {
+	return []string{
+		c.Car,
+		fmt.Sprintf("%f", c.MPG),
+		fmt.Sprintf("%d", c.Cylinders),
+		// remaining columns ...
+	}, nil
+}
+```
+
+This works, but adding, removing, or reordering any of the columns requires us to modify our code in 4 to 5 different places, which we wanted to avoid. So let's try some abstraction that allows us to declaratively specify our columns and how to marshal/unmarshal them:
```

Comment:
I'd add a reference to the paragraph in the beginning, where you mention the repeating-in-4-5-places. Otherwise it sounds a bit repetitive.

```suggestion
This works, but adding, removing, or reordering any of the columns requires us to modify our code in 4 to 5 different places, and as I mentioned in the beginning, we want to avoid that. So let's try some abstraction that allows us to declaratively specify our columns and how to marshal/unmarshal them:
```